src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs

```diff
@@ -726,5 +729,382 @@ void M(delegate*<delegate*<object, void>, void> param1, delegate*<delegate*<obje
+    public void M2()
+    {
+        delegate*<string> p1 = null;
+        delegate*<object> p2 = null;
+        M1(p1, p2);
+    }
+}");
+
+            // This should be inferrable with variant conversions, tracked by https://github.com/dotnet/roslyn/issues/39865
+            comp.VerifyDiagnostics(
+                // (9,9): error CS0411: The type arguments for method 'C.M1<T>(delegate*<T>, delegate*<T>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+                //         M1(p1, p2);
+                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "M1").WithArguments("C.M1<T>(delegate*<T>, delegate*<T>)").WithLocation(9, 9)
+            );
+        }
+
+        [Fact]
+        public void FunctionPointerGenericSubstitutionConflictingReturnSubstitution()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    public void M1<T>(delegate*<T> param1, delegate*<T> param2) {}
+    public void M2()
+    {
+        delegate*<string> p1 = null;
+        delegate*<int> p2 = null;
+        M1(p1, p2);
+    }
+}");
+
+            comp.VerifyDiagnostics(
+                // (9,9): error CS0411: The type arguments for method 'C.M1<T>(delegate*<T>, delegate*<T>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+                //         M1(p1, p2);
+                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "M1").WithArguments("C.M1<T>(delegate*<T>, delegate*<T>)").WithLocation(9, 9)
+            );
+        }
+
+        [Fact]
+        public void FunctionPointerGenericSubstitutionInference()
+        {
+
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    public void M1<T>(delegate*<T, T> param) {}
+    public void M2<T>()
+    {
+        delegate*<T, T> p = null;
+        M1(p);
+    }
+}");
+
+            comp.VerifyDiagnostics();
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var invocation = tree.GetRoot().DescendantNodes().OfType<InvocationExpressionSyntax>().Single();
+            var m1InvocationSymbol = (IMethodSymbol)model.GetSymbolInfo(invocation).Symbol!;
+
+            Assert.NotSame(m1InvocationSymbol, m1InvocationSymbol.OriginalDefinition);
+            Assert.Equal(TypeKind.TypeParameter, m1InvocationSymbol.TypeArguments[0].TypeKind);
+            var functionPointer = (IFunctionPointerTypeSymbol)m1InvocationSymbol.Parameters[0].Type;
+            Assert.Equal(TypeKind.TypeParameter, functionPointer.Signature.ReturnType.TypeKind);
+            Assert.Equal(TypeKind.TypeParameter, functionPointer.Signature.Parameters[0].Type.TypeKind);
+
+            var m2DeclarationSyntax = tree.GetRoot().DescendantNodes().OfType<MethodDeclarationSyntax>().Skip(1).Single();
+            var declaredSymbol = (IMethodSymbol)model.GetDeclaredSymbol(m2DeclarationSyntax)!;
```

Comment:
>var declaredSymbol = (IMethodSymbol)model.GetDeclaredSymbol(m2DeclarationSyntax)!; [](start = 12, length = 82)

This looks more complicated than it should be, we can locate the symbol from symbol table, there are a lot of helpers for that.